Home task for checking

NikitaTitov opened the pull request.

NikitaTitov committed:
Home task for checking

AlexanderShushunov commented:
Один коммит на все задание?

AlexanderShushunov commented:
И где тесты?

AlexanderShushunov commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Application {
+    public static void main(String args[]) {
+
```
Что за лишние пробелы?

NikitaTitov replied:
Это кажется, стандартное форматирование IDea

AlexanderShushunov commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,29 @@
+    public static void main(String args[]) {
+
+
+        try {
```
А все ли должно быть в try?

AlexanderShushunov commented on src/main/java/WorkingZone/MonitoredField.java:
```diff
@@ -0,0 +1,39 @@
+        monitoredField[row][column] = cellValue;
+    }
+
+    public StateForCell getStatusOfCell(int row, int column) {
```
так  State или Status?

AlexanderShushunov commented on src/main/java/WorkingZone/StateForCell.java:
```diff
@@ -0,0 +1,7 @@
+package WorkingZone;
+
+
+public enum StateForCell {
```
просто CellState

AlexanderShushunov commented on src/main/java/utils/cli/CLIParser.java:
```diff
@@ -0,0 +1,75 @@
+    public CLIParser() {
+        options = new Options();
+
+        Option leftOperand = Option.builder("h")
```
leftOperand ?

AlexanderShushunov commented on src/main/java/utils/cli/CLIParser.java:
```diff
@@ -0,0 +1,75 @@
+            printHelp();
+        }
+        line = parser.parse(options, args);
+        int height =  Integer.parseInt(line.getOptionValue("height"));
```
а если ParseException? не надо help печатать? И почему parser печатает?

AlexanderShushunov commented on src/main/java/WorkingZone/MonitoredField.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public MonitoredField clone() {
+        MonitoredField clone = new MonitoredField(getHeightOfField(), getWidthOfField());
```
для массивов есть стандартный DeepClone

NikitaTitov replied:
Я плохо гуглил, и там говорят что аборигены в библиотеках умеют копировать только одномерные массивы((

AlexanderShushunov commented on src/main/java/WorkingZone/utils/PrinterOfMonitoredField.java:
```diff
@@ -0,0 +1,34 @@
+
+    }
+
+    public static String print(MonitoredField field) {
```
Лучше без статики.

AlexanderShushunov commented on src/main/java/WorkingZone/utils/PrinterOfMonitoredField.java:
```diff
@@ -0,0 +1,34 @@
+import WorkingZone.MonitoredField;
+import WorkingZone.StateForCell;
+
+public class PrinterOfMonitoredField {
```
MonitoredFieldPrinter

AlexanderShushunov commented on src/main/java/utils/RandomFiller.java:
```diff
@@ -0,0 +1,20 @@
+    public static void fill(MonitoredField field, double fillFactor) {
+        for (int row = 0; row < field.getHeightOfField(); row++) {
+            for (int column = 0; column < field.getWidthOfField(); column++) {
+                if (Math.random() <= fillFactor) {
```
Может вынести в отдельную функцию получитьРандомныйСтатус?

AlexanderShushunov commented on src/main/java/utils/ReportPrinter.java:
```diff
@@ -0,0 +1,62 @@
+        this.groups = groups;
+    }
+
+    public String buildReport() {
```
👍

AlexanderShushunov commented on src/main/java/utils/DetectorOfGroups.java:
```diff
@@ -0,0 +1,56 @@
+        this.monitoredField = field.clone();
+    }
+
+    public HashMap<EmergencyStatus, Integer> detect() {
```
просто Map

AlexanderShushunov commented on src/main/java/utils/RiskDetector.java:
```diff
@@ -0,0 +1,16 @@
+
+
+public class RiskDetector {
+    private static double riskTable[] = {2, 4, 7, 13};
```
Нет явной связи между степенью риска и размером группы.

NikitaTitov committed:
Рефакторинг кода
Работа с названиями классов и методов
Изменено хранение данных в Map, с HashMap на TreeMap, для упорядоченного хранения и выведения ключей/значений
В классе перечислений статусов угроз создан конструктор, теперь перечисления хранят в себе нижние значения границ групп риска.
Добавлены методы в класс статусов.
  Добавлены тесты

AlexanderShushunov commented on src/main/java/WorkingZone/CellState.java:
```diff
@@ -0,0 +1,7 @@
+package WorkingZone;
```
маленькими буквами

AlexanderShushunov commented on src/main/java/WorkingZone/utils/MonitoredFieldPrinter.java:
```diff
@@ -0,0 +1,34 @@
+        for (int row = 0; row < field.getHeightOfField(); row++) {
+            result.append(row + "\t");
+            for (int column = 0; column < field.getWidthOfField(); column++) {
+                if (field.getCellState(row, column) == CellState.FILL)
```
Вынести в отдельную функцию + СКОБКИ!!!

AlexanderShushunov commented on src/main/java/utils/DetectorOfGroups.java:
```diff
@@ -0,0 +1,58 @@
+
+    public Map<EmergencyStatus, Integer> detect() {
+        Map<EmergencyStatus, Integer> groups = new TreeMap<EmergencyStatus, Integer>();
+        for (EmergencyStatus group : EmergencyStatus.values()) {
```
в функцию

AlexanderShushunov replied:
В приватную функци.
создали мапу
инициализаровали
заполнили

AlexanderShushunov commented on src/main/java/utils/DetectorOfGroups.java:
```diff
@@ -0,0 +1,58 @@
+
+                monitoredField.setCellValue(row, column, CellState.FREE);
+
+                groupSize += findGroupSize(row, column + 1);
```
в отдельный класс

AlexanderShushunov replied:
Тут напрашивается что-то типа класса "направление".

AlexanderShushunov commented on src/main/java/utils/EmergencyStatus.java:
```diff
@@ -0,0 +1,24 @@
+        return riskGrade;
+    }
+
+    public static int getSizeStatuses() {
```
лишний метод

AlexanderShushunov commented on src/main/java/utils/cli/IncomeData.java:
```diff
@@ -0,0 +1,26 @@
+
+
+public class IncomeData {
+    private int height;
```
final

AlexanderShushunov commented on src/test/java/TestEmergencySystem.java:
```diff
@@ -0,0 +1,52 @@
+
+public class TestEmergencySystem {
+    @Test
+    public void detect(){
```
маовато тестов

AlexanderShushunov commented on src/test/java/TestEmergencySystem.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Test
+    public void emergencyStatus(){
```
лишний

NikitaTitov committed:
Рефакторинг кода
Работа с названиями классов и методов
Убраны статические методы печати
Добавлены тесты для входной строки и детектора групп
Добавлены вспомогательные классы для тестов
Детектор групп теперь использует вернюю границу группы риска
Обновлены значения групп риска

AlexanderShushunov commented on src/main/java/Application.java:
```diff
@@ -11,18 +11,20 @@
-            System.out.println(reportPrinter.buildReport());
+            incomeData = parser.parse(args);
         } catch (ParseException e) {
             System.out.println("ParseException in main class");
```
и что дальше...
Вообще пользователь любит такие подробные сообщения об ошибках.

NikitaTitov replied:
А как сделать правильно? У меня метод parse кидает ParseException внутри класса CLIParser. Exception выдаст стактрэйс и затем напечатает "ParseException in main class". Правильно я понимаю?

AlexanderShushunov replied:
он упадет не здесь

AlexanderShushunov replied:
А хелп вывести, а программу завершить.

NikitaTitov replied:
Help выведет парсер, а закрыть программу действительно нужно((

AlexanderShushunov commented on src/main/java/utils/cli/HelpPrinter.java:
```diff
@@ -0,0 +1,14 @@
+import org.apache.commons.cli.Options;
+
+
+public class HelpPrinter {
```
Зачем в отдельный класс?

NikitaTitov replied:
Лишний метод в классе CLIParser, не относящийся к классу, поэтому решил вынести

AlexanderShushunov replied:
не лишний. Этот класс очень сильно связан с CLIParser и без него не имеет смысл.

NikitaTitov replied:
Ок внесу обратно, но сильной связи нет. Класс парсер парсит строку и это его единственное предназначение. если у него что то не получается он либо кидает exception либо просит другой класс вывести на печать подсказку - помощь
